### ISSUE TITLE
Provide documentation for running on k8s with run-config based flow

### ISSUE THREAD

**Issue by Zaubeerer** [PrefectHQ/prefect#3510]
## Description
As described in https://github.com/PrefectHQ/prefect/issues/3509, we have problems enabling debug level logging for our `DaskKubernetesEnvironment`.

Therefore, we tried out `flow.run_config`, which was successful.

However, it seems like tasks are not parallelized properly.

We tried both `flow.executor = DaskExecutor()` and `flow.executor = LocalDaskExecutor(num_workers=...)`, but got the following gantt charts:

flow.run_config :

![image](https://user-images.githubusercontent.com/60844432/96130164-d5b43480-0ef7-11eb-8e3d-5923bb0b7825.png)

flow.run_config - `flow.executor = DaskExecutor()`  :
![image](https://user-images.githubusercontent.com/60844432/96131310-4e67c080-0ef9-11eb-973a-d32c80c79e63.png)

flow.run_config - `flow.executor = LocalDaskExecutor(num_workers=20)` :

![image](https://user-images.githubusercontent.com/60844432/96131814-f4b3c600-0ef9-11eb-87a7-f9c7c76a158e.png)

Although the tasks do parallelize for the LocalDaskExecutor, they take 20 times as long (36 s instead of 1-2 seconds).
## Expected Behavior
- screenshot of parallelized run (albeit without debug level logging)

DaskKubernetesEnvironment:
![image](https://user-images.githubusercontent.com/60844432/96129816-64748180-0ef7-11eb-8ef9-600d7221c75c.png)

## Reproduction

- add parallelization to minimal example

## Environment
- AWS EKS cluster 

<details>

```python
## KUBERNETES
prefect_agent_labels = {
	"app": "prefect-agent",
}

prefect_agent_deployment = k8s.apps.v1.Deployment(
	"prefect-agent",
	spec={
		"replicas": 1,
		"selector": {
			"match_labels": prefect_agent_labels
		},
		"template": {
			"metadata": {
				"labels": prefect_agent_labels
			},
			"spec": {
				"containers": [{
					"name": "agent",
					"resources": {
						"limits": {
							"cpu": "100m",
							"memory": "128Mi",
						}
					},
					"args": [
						"prefect agent start kubernetes",
					],
					"command":[
						"/bin/bash", "-c"
					],
					"env":[
						{"name": "PREFECT__CLOUD__AGENT__AUTH_TOKEN",
                        "value": PREFECT_AGENT_AUTH},
						{"name": "PREFECT__CLOUD__API",
						 "value": "https://api.prefect.io", },
						{"name": "NAMESPACE",
						 "value": "default", },
						{"name": "IMAGE_PULL_SECRETS",
						 "value": "", },
						{"name": "PREFECT__CLOUD__AGENT__LABELS",
						 "value": PREFECT_AGENT_LABELS, },
						{"name": "PREFECT__LOGGING__LEVEL",
						 "value": "DEBUG"},
						{"name": "PREFECT__CLOUD__LOGGING__LEVEL",
						 "value": "DEBUG"},
						{"name": "JOB_MEM_REQUEST",
						 "value": "", },
						{"name": "JOB_MEM_LIMIT",
						 "value": "", },
						{"name": "JOB_CPU_REQUEST",
						 "value": "", },
						{"name": "JOB_CPU_LIMIT",
						 "value": "", },
						{"name": "IMAGE_PULL_POLICY",
						 "value": "", },
						{"name": "SERVICE_ACCOUNT_NAME",
						 "value": "", },
						{"name": "PREFECT__BACKEND",
						 "value": "cloud", },
						{"name": "PREFECT__CLOUD__AGENT__AGENT_ADDRESS",
						 "value": "http://:8080", },
					],
					"image": PREFECT_AGENT_IMAGE,
					#"image": "prefecthq/prefect:0.12.3-python3.6",
					"image_pull_policy": "Always",
					"liveness_probe": {
						"failure_threshold": 2,
						"http_get": {
							"path": "/api/health",
							"port": 8080
						},
						"initial_delay_seconds": 40,
						"period_seconds": 40,
					},
				}]
			}
		}
	},
	opts=pulumi.ResourceOptions(provider=k8s_provider)
)
```

</details>

Optionally run `prefect diagnostics` from the command line and paste the information here. -->

**Comment by Zaubeerer** [comment]
Maybe there is only some fine tuning with the dask parameters required?

However, it would be great to set the default parameters such, that parallelization works out-of-the-box to simplify user experience.

**Comment by Zaubeerer** [comment]
Tweaked the parameters as follows, without success:

```python
run_config = KubernetesRun(cpu_limit=2, cpu_request=2)
executor = LocalDaskExecutor(num_workers=4)
```


![image](https://user-images.githubusercontent.com/60844432/96160823-bf69a100-0f16-11eb-8f0b-624bd66b58df.png)

**Comment by Zaubeerer** [comment]
I just tried to add `n_workers` as kwarg to DaskExecutor, but it does not seem to be successful, as the flow does not start...

```python
flow.run_config = KubernetesRun(
    env={"PREFECT__LOGGING__LEVEL": "DEBUG"},
    labels=["k8s"],
)

flow.executor = DaskExecutor(
    cluster_kwargs={
            "n_workers": 20,
        },)
```


![image](https://user-images.githubusercontent.com/60844432/96171406-27bf7f00-0f25-11eb-90e8-613c66d65423.png)

**Comment by jcrist** [comment]
Hi @Zaubeerer, sorry to hear you're having issues. The `run_config` stuff is still experimental (and undocumented), so thanks for trying things out. I believe all the issues you're having have to do with your prefect/k8s configuration, and aren't evidence of a bug in prefect/dask/something else. Responding inline:

> We tried both `flow.executor = DaskExecutor()` and `flow.executor = LocalDaskExecutor(num_workers=...)`, but got the following gantt charts:

From the community slack, it looks like you're used to using the `DaskKubernetesEnvironment`, which runs your flow spread out across a number of k8s pods. Switching to using `KubernetesRun` will have your flow run using a single k8s pod (without additional executor configuration) - depending on your flow this may be fine or lead to perf issues. By default the flow will run serially, one task at a time, but (as you've rightly noted) you can set an `executor` on the flow to re-enable parallelism. 

Both options you've tried will run within the same pod using either threads or processes - but for this to be effective, the pod needs sufficient resources to run in parallel. For IO heavy tasks, using more threads than you have cores allocated to your pod is fine, but for compute heavy tasks you don't want more processes/threads than you have cores (I'm not sure what your flow is doing, but it sounds like it's compute heavy).

In the latest release (0.13.11) there are no default resource limits for a job, your job will be scheduled wherever the k8s cluster sees fit. In previous releases, the cpu request/limit were limited to < 0.1 cpu, which isn't enough for any compute-heavy task. You can always set your own limit/request (as you've found) by passing `cpu_limit`/`cpu_request` to `KubernetesRun`.

> With the following code, the flow is not even started. Is there a contradiction of what the flow asks for (5 CPUs) and what the AWS EKS is configured for?

Yeah, if you request a pod that requires more resources than any node in the cluster can give, it will hang until there's a place for it (possibly forever). This depends on the node configurations of your system, and if you have any auto-scaling node pools. Running everything in a single pod is simpler, and is what I'd recommend when possible, but once your compute gets large enough it may be less expensive to scale out to multiple pods than require a larger cluster node be available.

> I just tried to add n_workers as kwarg to DaskExecutor, but it does not seem to be successful, as the flow does not start...

From that configuration, I'd guess you're swamping the node, as it doesn't have enough cpu to manage that many workers. The flow would eventually run, but it would be so slow because everything is highly oversubscribed.

---

I believe the above explains the issues you've been seeing, please let me know if you have any questions.

I plan to write a bunch of example docs for deployments with run configs, and there will certainly be a large section on k8s (both with and without dask), but for now there's unfortunately not anything I can point you to. To use dask-kubernetes without the `DaskKubernetesEnvironment`, you'd want to configure your `DaskExecutor` with a `cluster_class`/`cluster_kwargs`/`adapt_kwargs` to start and scale a `dask_kubernetes.KubeCluster` appropriately.

- `cluster_class`: either the import name of a callable, or a callable to create a dask cluster. Ideally `"dask_kubernetes.KubeCluster"` would be all you'd need, but not all options you might want to change (like the worker images) are exposed in the constructor without writing k8s specs. This is unfortunate, I'm hoping to upstream some patches to make this easier to do.
- `cluster_kwargs`: kwargs to pass to `cluster_class` at runtime to create a cluster.
- `adapt_kwargs`: kwargs to pass to `cluster.adapt` at runtime to enable adaptive scaling, if desired.

Ideally I'd like the following to work:

```python
import dask_kubernetes

flow.executor = DaskExecutor(
    cluster_class="dask_kubernetes.KubeCluster",
    cluster_kwargs={"image": "your-worker-image-here", "n_workers": 4}
)
```

But `image` isn't exposed in the `KubeCluster` constructor, so you'd need to do something more like:

```python
import dask_kubernetes

def make_cluster():
    """A function that makes a dask-kubernetes cluster with the configuration you want"""
    spec = dask_kubernetes.make_pod_spec(
        image="your-worker-image",
        ...
    )
    cluster = dask_kubernetes.KubeCluster(spec, n_workers=4)
    return cluster

flow.executor = DaskExecutor(cluster_class=make_cluster)

# if you wanted adaptive scaling instead of fixed scaling, you'd drop the
# `n_workers` in `make_cluster` and do something like:
# 
# flow.executor = DaskExecutor(cluster_class=make_cluster, adapt_kwargs={"maximum": 10})
```

(you might find the [dask-kubernetes docs](https://kubernetes.dask.org/en/latest/kubecluster.html) useful here).

Apologies that there aren't better docs on this - as I said before you're using some very new features. If you do try out the above, please let us know how things work for you.

**Comment by Zaubeerer** [comment]
Dear @jcrist, 

thanks a lot for the broad and deep feedback!

As #3509 got fixed with yesterday's release, this issue became less urgent for us.

However, I would be happy to contribute and help testing `run_config`, e.g. by giving feedback for the examples from a "new user" perspective :)

**Comment by jcrist** [comment]
Definitely! I'll ping you once we have some docs up. I'm going to change this issue to a docs issue for running on k8s run config, since that'll be mostly taking the above info and converting to docs :).

**Comment by jcrist** [comment]
With the 0.14.0 release last week, `flow.run_config` is now the default flow-configuration system, with `flow.environment` being deprecated. The docs have been overhauled to (hopefully) provide a smoother experience. See:

- https://docs.prefect.io/orchestration/tutorial/overview.html
- https://docs.prefect.io/orchestration/flow_config/overview.html
- https://docs.prefect.io/orchestration/agents/overview.html

We definitely plan on expanding documentation further, but hopefully this is enough to get you started. Closing for now, feel free to open a new issue if you find certain parts of the docs lacking.
